packaging/bin/create-deb-repo.sh

```diff
@@ -0,0 +1,116 @@
+#!/bin/bash
+set -eEuo pipefail
+
+# External dependencies:
+# - https://salsa.debian.org/apt-team/apt (apt-ftparchive, packaged in apt-utils)
+# - https://aws.amazon.com/cli/
+#   awscli expects AWS_ACCESS_KEY_ID and AWS_SECRET_ACCESS_KEY to be set in the
+#   environment.
+# - https://gnupg.org/
+#   For signing the script expects the private signing key to already be
+#   imported and PGPKEYID and PGP_SIGN_KEY_PASSPHRASE to be set in the
+#   environment.
+# - generate_index.py
+#   For generating the index.html of each directory. It's available in the
+#   packaging/bin directory of the k6 repo, and should be in $PATH.
+
+_s3bucket="${S3_BUCKET-dl.k6.io}"
+_usage="Usage: $0 <pkgdir> <repodir> [s3bucket=${_s3bucket}]"
+PKGDIR="${1?${_usage}}"  # The directory where .deb files are located
+REPODIR="${2?${_usage}}" # The package repository working directory
+S3PATH="${3-${_s3bucket}}/deb"
+# Remove packages older than N number of days (730 is roughly ~2 years).
+REMOVE_PKG_DAYS=730
+
+log() {
+  echo "$(date -Iseconds) $*"
+}
+
+delete_old_pkgs() {
+  find "$1" -name '*.deb' -type f -daystart -mtime "+${REMOVE_PKG_DAYS}" -print0 | xargs -r0 rm -v
+
+  # Remove any dangling .asc files
+  find "$1" -name '*.asc' -type f -print0 | while read -r -d $'\0' f; do
+    if ! [ -r "${f%.*}" ]; then
+      rm -v "$f"
+    fi
+  done
+}
+
+sync_to_s3() {
+  log "Syncing to S3 ..."
+  aws s3 sync --no-progress --delete "${REPODIR}/" "s3://${S3PATH}/"
+
+  # Set a short cache expiration for index and repo metadata files.
+  aws s3 cp --no-progress --recursive \
+    --exclude='*.deb' --exclude='*.asc' --exclude='*.html' \
+    --cache-control='max-age=60,must-revalidate' \
+    --metadata-directive=REPLACE \
+    "s3://${S3PATH}" "s3://${S3PATH}"
+  # Set it separately for HTML files to set the correct Content-Type.
+  aws s3 cp --no-progress --recursive \
+    --exclude='*' --include='*.html' \
+    --content-type='text/html' \
+    --cache-control='max-age=60,must-revalidate' \
+    --metadata-directive=REPLACE \
```

Comment:
what does `Note that if the object is copied over in parts,` from the documentation of this option means ?

I can't find an explanation so I wonder if this is just something that `aws s3 cp` won't do on it's own? or is it something that has a flag that is just badly documented and me searching ... fails

Reply:
Yeah, the documentation is a bit confusing... I think this is mostly relevant when doing multipart uploads, but since we're copying from the same bucket and just replacing the metadata, I don't think it matters. I did have to specify `content-type` manually because otherwise everything gets `application/octet-stream` :-/

This was more intuitive with `s3cmd` which had a `modify` command, so hopefully this copying over doesn't incur additional costs.